/*  See the NOTICE file distributed with this work for additional information
 *  regarding copyright ownership.
 *
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License. */

syntax = "proto3";

// Client-facing representation of a core ValueSet item (from CDM) exposed in APIs.
message ValueSetItem {
    
    // Required field identifying the ValueSet
    // This has the format `topic`.`unique_identifier` - e.g. mane.select.1
    string accession_id = 1;
    
    // Required field containing a short human readable name for the ValueSet item, 
    // which could be used for display purposes - e.g. "Mane Select"
    optional string label = 2;

    // Optional field containing a short machine readable name for the ValueSet item. 
    // This should be lowercase and snake_case - e.g. "protein_coding_gene"
    optional string value = 3; 

    // Optional field containing a short description of the ValueSet item
    optional string definition = 4;

    // Optional field containing a more verbose definition of the ValueSet item
    optional string description = 5;

    // Optional field containing a flag to state whether ValueSet is current or not
    optional bool is_current = 6;
}


message ValueSetRequest {
    // What ValueSet to read. If the viewer doesn't have access to the ValueSet or the
    // ValueSet has been deleted, the response will be empty but will succeed.
    optional string accession_id   = 1;
    optional bool   use_noncurrent = 2;
}


service ValueSet {
    rpc GetValueSetByAccessionId (ValueSetRequest) returns (ValueSetItem);
    rpc GetValueSetsByTopic (ValueSetRequest) returns (stream ValueSetItem);
    rpc GetAllValueSets (ValueSetRequest) returns (stream ValueSetItem);
}
